cpp/src/fil/infer.cu

```diff
@@ -96,14 +96,14 @@ __device__ __forceinline__ vec<NITEMS, output_type> infer_one_tree(
   tree_type tree, const float* input, int cols, int n_rows) {
   int curr[NITEMS];
   // the first n_rows are active
-  int mask = ((1 << n_rows) - 1) << (NITEMS - n_rows);
+  int mask = (1 << n_rows) - 1;
   for (int j = 0; j < NITEMS; ++j) curr[j] = 0;
   do {
 #pragma unroll
     for (int j = 0; j < NITEMS; ++j) {
       auto n = tree[curr[j]];
       mask &= ~(n.is_leaf() << j);
-      if (!n.is_leaf()) {
+      if (mask & (1 << j)) {
```

Comment:
fixing a bug which somehow did not trigger the test in the previous PR. It triggers it now.

Reply:
Which test catches this? Does this indicate we need a tighter test here if the previous one did not pick up the bug?

Reply:
The tests where number of rows % 4 != 0 and algo = BATCH_TREE_REORG, below.
The only test that would reliably catch this is the set below + compute-sanitizer [memcheck], which would be much slower than a normal FIL test. Do you have any ideas?

FIL_TEST_PARAMS(num_rows = 101, num_cols = 100'000, depth = 5, num_trees = 9,
   algo = BATCH_TREE_REORG, leaf_algo = GROVE_PER_CLASS,
   num_classes = 3),
     FIL_TEST_PARAMS(num_rows = 102, num_cols = 100'000, depth = 5,
                  num_trees = FIL_TPB + 1, algo = BATCH_TREE_REORG,
                  num_trees = 3 * (FIL_TPB + 1), algo = BATCH_TREE_REORG,
                            leaf_algo = GROVE_PER_CLASS, num_classes = FIL_TPB + 1),
  FIL_TEST_PARAMS(num_rows = 103, num_cols = 100'000, depth = 5, num_trees = 1,
                  algo = BATCH_TREE_REORG, leaf_algo = CATEGORICAL_LEAF,
                    num_classes = 3),